In [1]:
#Importing dependencies 
import pandas as pd
import numpy as np

In [2]:
#Getting a file path 
file_path = "purchase_data.csv"

#Reading data into a dataframe 
data = pd.read_csv(file_path)

In [3]:
#Getting a snapchat of data
data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [4]:
#Getting a snapchat of data
data.shape

(780, 7)

In [5]:
#Display the total number of players
players = data["SN"].nunique()
print(f"The total number of unique players: {players}")


The total number of unique players: 576


In [6]:
#Purchasing Analysis (Total)

#Number of Unique Items
items = data["Item ID"].nunique()
print(f"The total number of unique items: {items}")

#Average Purchase Price
avr_price = round(data["Price"].sum()/data["Price"].count(),2)
print(f"The average price of an item: {avr_price}")

#Total Number of Purchases
purchases = data["Purchase ID"].count()
print(f"The total number of purchases: {purchases}")

#Total Revenue
revenue = data["Price"].sum()
print(f"Total revenue: {revenue}") 


The total number of unique items: 183
The average price of an item: 3.05
The total number of purchases: 780
Total revenue: 2379.77


In [7]:
#Putting the purchasing summary into a dataframe

d = {'Number of unique items': [items], 'Total number of purchases': [purchases], 'Average price': [avr_price], 'Total revenue': [revenue]}
summary_df = pd.DataFrame(data=d)

#Adding dollar signs to dollar values 
summary_df[['Average price', 'Total revenue']] = summary_df[['Average price', 'Total revenue']].applymap('${:,.2f}'.format)
summary_df

Number of unique items  Total number of purchases Average price  \
0                     183                        780         $3.05   

  Total revenue  
0     $2,379.77

In [8]:
#Gender Demographics

#Count all players that made purchases
total = data['SN'].count()
print(f"There are {total} players in total")

#Count all unique players
unique_total = data['SN'].nunique()
print(f"However, there are {unique_total} unique players")  

#Count of unique players in each gender category:

m_unique= data.groupby('Gender')['SN'].nunique()['Male']
print(f"There are {m_unique} male players")

f_unique = data.groupby('Gender')['SN'].nunique()['Female']
print(f"There are {f_unique} female players")

other_unique = data.groupby('Gender') ['SN'].nunique()["Other / Non-Disclosed"]
print(f"There are {other_unique} other players")


#Count percentages of genders:
m_percent = round((m_unique*100)/(unique_total), 2)
print(f"{m_percent}% of all players are male")

f_percent = round((f_unique*100)/(unique_total), 2)
print(f"{f_percent}% of all players are female")

other_percent = round((other_unique*100)/(unique_total), 2)
print(f"{other_percent}% of all players are other")




There are 780 players in total
However, there are 576 unique players
There are 484 male players
There are 81 female players
There are 11 other players
84.03% of all players are male
14.06% of all players are female
1.91% of all players are other


In [9]:
#Another way to see unique players of each gender category  
players= data.groupby('Gender')['SN'].nunique()
players = players.sort_values(ascending=False).tolist()
players



[484, 81, 11]

In [10]:
#Putting the demographics summary into a dataframe
columns = {'Percentage of players': [m_percent, f_percent, other_percent], 'Total Count' : [m_unique, f_unique, other_unique]}
demographics_df = pd.DataFrame(data = columns, index = ["Male", "Female", "Other"])
demographics_df['Percentage of players'] = demographics_df['Percentage of players'].map('{:,.2f}%'.format)
demographics_df



Percentage of players  Total Count
Male                  84.03%          484
Female                14.06%           81
Other                  1.91%           11

In [61]:
#Purchasing Analysis:

#Purchase count (total):
purchase_count = data['Purchase ID'].count()
print (f"There has been {purchase_count} purchases made")

#Average purchase price per person:
avr_purchase = round(revenue/unique_total, 2)
print (f"The average purchase price per person is {avr_purchase}$")

#Total revenue per gender:
total_per_gender = data.groupby('Gender')['Price'].sum()
total_per_gender = total_per_gender.sort_values(ascending=False)
total_per_gender = total_per_gender.tolist()

#Purchase count (per gender):
count_gender = data.groupby('Gender')['SN'].count()
count_gender = count_gender.sort_values(ascending=False)
count_gender = count_gender.tolist()

#Avgerage purchase price per gender:

gender_average = []
for x, y in zip(total_per_gender, count_gender) :
        gender_average.append (round(x/y, 2))
        print(gender_average)

#Avgerage purchase price per gender per person: 

person_average = []
for i,j in zip(total_per_gender, players):
    person_average.append (round(i/j, 2))
    print(person_average)

purchasing_data = {"Purchase count": count_gender, "Average purchase price": gender_average, "Total Purchase Value": total_per_gender, "Average purchase price per person":person_average}
purchasing_df = pd.DataFrame(data = purchasing_data, index = ["Male", "Female", "Other"])

#Let's see how much profit each gender generated:

male_profit = purchasing_df['Total Purchase Value']['Male']
female_profit = purchasing_df['Total Purchase Value']['Female']

male_percentage = round((male_profit*100)/revenue,2)
female_percentage = round((female_profit*100)/revenue,2)
print(f"Male players have generated {male_percentage}% of profit")
print(f"Female players have generated {female_percentage}% of profit")


#Displaying a dataframe 
purchasing_df[['Average purchase price', 'Total Purchase Value', 'Average purchase price per person']] = purchasing_df[['Average purchase price', 'Total Purchase Value', 'Average purchase price per person']].applymap('${:,.2f}'.format)
purchasing_df





There has been 780 purchases made
The average purchase price per person is 4.13$
[3.02]
[3.02, 3.2]
[3.02, 3.2, 3.35]
[4.07]
[4.07, 4.47]
[4.07, 4.47, 4.56]
Male players have generated 82.68% of profit
Female players have generated 15.21% of profit


Purchase count Average purchase price Total Purchase Value  \
Male               652                  $3.02            $1,967.64   
Female             113                  $3.20              $361.94   
Other               15                  $3.35               $50.19   

       Average purchase price per person  
Male                               $4.07  
Female                             $4.47  
Other                              $4.56

In [12]:
#Age Demographics:

#Establish bins for ages
age_bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

#Categorize the existing players using the age bins:
data["Age group"] = pd.cut(data["Age"], age_bins, labels=group_names)

#Calculate the numbers and percentages by age group
group_sizes = data.groupby('Age group')['SN'].nunique()
group_distributions = round(((group_sizes*100)/unique_total),2).map('{:,.2f}%'.format)
group_distributions = pd.DataFrame(group_distributions)

#Create a summary data frame to hold the results
age_demographics = group_distributions.merge(group_sizes.to_frame(), left_index=True, right_index=True)
age_demographics =  age_demographics.rename(columns={'SN_x' : 'Percentage of players', 'SN_y':'Total Count'})
age_demographics



Percentage of players  Total Count
Age group                                   
<10                       2.95%           17
10-14                     3.82%           22
15-19                    18.58%          107
20-24                    44.79%          258
25-29                    13.37%           77
30-34                     9.03%           52
35-39                     5.38%           31
40+                       2.08%           12

In [13]:
#Purchasing Analysis (Age)

#Bin the purchase_data data frame by age
data_1 = data.groupby('Age group').count()
data_1

#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below

aggregations = {'Purchase ID': {'Purchase Count':'count'},'Price': {'Average purchase price': 'mean','Total Purchase Value':'sum'}}

data_1 = data.groupby(['Age group']).agg(aggregations)

#Count average purchase price per person per group
unique_players = data.groupby('Age group')['SN'].nunique()
purchase_total = data.groupby('Age group')['Price'].sum()

avg_per_person = []
for a,b in zip(unique_players, purchase_total):
    avg_per_person.append(round((b/a),2))

avg_per_person = pd.Series(avg_per_person)

data_1.columns = data_1.columns.droplevel()    
data_1['Average total purchase per person'] = avg_per_person.values    

data_1[['Average purchase price', 'Average total purchase per person', 'Total Purchase Value']] = data_1[['Average purchase price', 'Average total purchase per person', 'Total Purchase Value']].round(2).applymap('${:,.2f}'.format)
data_1




/anaconda3/envs/PythonData/lib/python3.6/site-packages/pandas/core/groupby/groupby.py:4658: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


Purchase Count Average purchase price Total Purchase Value  \
Age group                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Average total purchase per person  
Age group                                    
<10                                   $4.54  
10-14                                 $3.76  
15-19                                 $3.86  
20-24                                 $4.32  
25-29                                 $3.81  
30-34                                 $4.12  
35-39                                 $4.76  
40+                                   $3.19

In [64]:
#Top Spenders:
#Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
g = {'Item ID': {'Purchase Count':'count'}, 'Price':{'Average purchase price': 'mean','Total Purchase Value':'sum'}}
top_spenders = data.groupby('SN')['Price'].agg(g)
top_spenders.columns = top_spenders.columns.droplevel()
top_spenders = top_spenders.sort_values(ascending=False, by = 'Total Purchase Value')[:5]
top_spenders[['Average purchase price','Total Purchase Value']] = top_spenders[['Average purchase price','Total Purchase Value']].applymap('${:,.2f}'.format)
top_spenders

/anaconda3/envs/PythonData/lib/python3.6/site-packages/ipykernel/__main__.py:4: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version
/anaconda3/envs/PythonData/lib/python3.6/site-packages/pandas/core/groupby/groupby.py:3558: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version
  results[name] = obj.aggregate(func)


Purchase Count Average purchase price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

In [16]:
f= {'Item ID': {'Purchase Count':'count'}, 'Price': {'Total Purchase Value': 'sum'}}
popular_items = data.groupby(['Item ID', 'Item Name', 'Price']).agg(f)
popular_items.columns = popular_items.columns.droplevel()
popular_items = popular_items.sort_values(ascending=False, by = 'Purchase Count')[:5]
popular_items = popular_items.reset_index(['Price'])
popular_items[['Price','Total Purchase Value']] = popular_items[['Price','Total Purchase Value']].applymap('${:,.2f}'.format)
popular_items

     

/anaconda3/envs/PythonData/lib/python3.6/site-packages/pandas/core/groupby/groupby.py:4658: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


Price  Purchase Count  \
Item ID Item Name                                                             
178     Oathbreaker, Last Hope of the Breaking Storm  $4.23              12   
145     Fiery Glass Crusader                          $4.58               9   
108     Extraction, Quickblade Of Trembling Hands     $3.53               9   
82      Nirvana                                       $4.90               9   
19      Pursuit, Cudgel of Necromancy                 $1.02               8   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

In [37]:
#Most Profitable Items
#Identify the 5 most profitable items by total purchase value
#Then list (in a table): Item ID, Item Name, Purchase Count, Item Price, Total Purchase Value
a = {'Price': {'Total Purchase Value': 'sum', 'Item price':'mean'}, 'Item ID': {'Purchase Count':'count'}}
profit_items = data.groupby(['Item ID', 'Item Name']).agg(a)
profit_items.columns = profit_items.columns.droplevel()
profit_items = profit_items.sort_values(ascending=False, by = 'Total Purchase Value')[:5]
profit_items[['Item price','Total Purchase Value']] = profit_items[['Item price','Total Purchase Value']].applymap('${:,.2f}'.format)
profit_items

/anaconda3/envs/PythonData/lib/python3.6/site-packages/pandas/core/groupby/groupby.py:4658: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


Total Purchase Value  \
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76   
82      Nirvana                                                    $44.10   
145     Fiery Glass Crusader                                       $41.22   
92      Final Critic                                               $39.04   
103     Singed Scalpel                                             $34.80   

                                                     Item price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                      Purchase Count  
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm              12  
82      Nirvana                                                    9  
145     Fiery Glass Crusader                                       9  
92      Final Critic                                               8  
103     Singed Scalpel                                             8

## Report:

* 1. The most of the revenue is generated by male players (82.68%), while female players generated other 15.21% of profit. 

* 2. In terms of age demographics, the most of the revenue is generated by the players that are between 20-24 years old (\$ 1,114.06 out of \$ 2379.77 or 46.8%)

* 3. Even the top spenders of the game do not spend more than $ 19 on additional game items. 

